src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,156 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "kudu/consensus/routing.h"
+
+#include <glog/logging.h>
+
+#include "kudu/gutil/map-util.h"
+#include "kudu/util/status.h"
+
+using std::string;
+using std::unique_ptr;
+using std::unordered_map;
+using std::vector;
+
+namespace kudu {
+namespace consensus {
+
+Status RoutingTable::Init(RaftConfigPB& config,
+                          const std::string& leader_uuid) {
+  unordered_map<string, Node*> index;
+  unordered_map<string, unique_ptr<Node>> trees;
+
+  RETURN_NOT_OK(ConstructForest(config, &index, &trees));
+  RETURN_NOT_OK(MergeTrees(leader_uuid, index, &trees));
+  ConstructNextHopIndicesRec(trees.begin()->second.get());
+
+  index_ = std::move(index);
+  topology_root_ = std::move(trees.begin()->second);
+
+  return Status::OK();
+}
+
+Status RoutingTable::ConstructForest(
+    RaftConfigPB& config,
+    std::unordered_map<std::string, Node*>* index,
+    std::unordered_map<std::string, std::unique_ptr<Node>>* trees) {
+
+  // Construct the peers.
+  for (const RaftPeerPB& peer : config.peers()) {
+    unique_ptr<Node> hop(new Node(peer));
+    index->emplace(peer.permanent_uuid(), hop.get());
+    auto result = trees->emplace(peer.permanent_uuid(), std::move(hop));
+    if (!result.second) {
+      return Status::InvalidArgument("invalid config: duplicate uuid",
+                                     peer.permanent_uuid());
+    }
+  }
+  // Organize the peers into a forest of trees.
+  for (const RaftPeerPB& peer : config.peers()) {
+    if (peer.proxy_from().empty()) {
+      continue;
+    }
+
+    // Node has a parent so we must link them and assign ownership to the
+    // parent.
+    const string& parent_uuid = peer.proxy_from();
+    Node* parent_ptr = FindWithDefault(*index, parent_uuid, nullptr);
+    if (!parent_ptr) {
+      return Status::InvalidArgument("invalid config: cannot find proxy",
+                                     parent_uuid);
+    }
+
+    // Move out of trees map and into parent as a child.
+    const string& node_uuid = peer.permanent_uuid();
+    auto iter = trees->find(node_uuid);
+    DCHECK(iter != trees->end());
+    unique_ptr<Node> node = std::move(iter->second);
+    trees->erase(iter->first);
+    node->parent = parent_ptr;
+    auto result = parent_ptr->children.emplace(node_uuid, std::move(node));
+    if (!result.second) {
+      return Status::InvalidArgument("invalid config: duplicate uuid",
+                                     node_uuid);
+    }
+  }
+  return Status::OK();
+}
+
+Status RoutingTable::MergeTrees(
```

Comment:
this function is missing comments.

Reply:
Yes, it needs them. Fixed. Don't forget to look at the header file comments though.

Reply:
Added some comments.